ducktape/tests/runner_client.py

```diff
@@ -0,0 +1,283 @@
+# Copyright 2016 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import time
+import traceback
+import zmq
+
+from ducktape.tests.event import ClientEventSupplier
+from ducktape.tests.loader import TestLoader
+from ducktape.tests.serde import SerDe
+from ducktape.tests.test import test_logger
+
+from ducktape.tests.result import TestResult, IGNORE, PASS, FAIL
+from ducktape.tests.reporter import SingleResultFileReporter
+from ducktape.utils.local_filesystem_utils import mkdir_p
+
+
+def run_client(server_hostname, server_port, logger_name, test_id, log_dir, debug, max_parallel):
+    client = RunnerClient(server_hostname, server_port, test_id, logger_name, log_dir, debug, max_parallel)
+    client.run()
+
+
+class RunnerClient(object):
+    """Run a single test"""
+
+    def __init__(self, server_hostname, server_port, test_id, logger_name, log_dir, debug, max_parallel):
+        self.serde = SerDe()
+        self.logger = test_logger(logger_name, log_dir, debug, max_parallel)
+        self.runner_port = server_port
+
+        self.test_id = test_id
+        self.id = "test-runner-%d-%d" % (os.getpid(), id(self))
+        self.message_supplier = ClientEventSupplier(self.test_id, self.id)
+        self.sender = Sender(server_hostname, str(self.runner_port), self.message_supplier, self.logger)
+
+        ready_reply = self.sender.send(self.message_supplier.ready())
+        self.session_context = ready_reply["session_context"]
+        self.test_metadata = ready_reply["test_metadata"]
+        self.cluster = ready_reply["cluster"]
+
+        # Wait to instantiate the test object until running the test
+        self.test = None
+        self.test_context = None
+
+    def send(self, event):
+        return self.sender.send(event)
+
+    def collect_test_context(self, directory, file_name, cls_name, method_name, injected_args):
```

Comment:
should probably be private. probably same for `send` above too, as well as some other methods below -- almost none of this class needs to be public. (would probably be good to start following good practice naming private/public fields as well)